# Task
Perform the following tasks on the dataset provided in the file "/content/IMDB Dataset.csv": 1. Apply necessary preprocessing techniques. 2. Find the total number of words and unique words in the corpus using Python. 3. Apply One-Hot Encoding. 4. Apply Bag of Words and find the vocabulary and word counts. 5. Apply Bag of Bigrams and Trigrams and observe the dimensionality of the vocabulary. 6. Apply TF-IDF, calculate IDF scores, and find the vocabulary.

## Load the dataset

### Subtask:
Load the dataset from "/content/IMDB Dataset.csv" into a pandas DataFrame.


**Reasoning**:
Load the dataset into a pandas DataFrame and display the first few rows and the data types.



In [2]:
df = pd.read_csv('/content/IMDB Dataset.csv', engine='python', on_bad_lines='skip')
display(df.head())
display(df.info())

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43525 entries, 0 to 43524
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     43525 non-null  object
 1   sentiment  43525 non-null  object
dtypes: object(2)
memory usage: 680.2+ KB


None

## Preprocessing

### Subtask:
Apply necessary preprocessing techniques such as removing HTML tags, punctuation, special characters, converting text to lowercase, and removing stop words.


**Reasoning**:
Define a function to preprocess text and apply it to the 'review' column to create a new 'preprocessed_review' column, then display the first few rows.



In [3]:
import re
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

def preprocess_text(text):
    # Remove HTML tags
    text = re.sub(r'<.*?>', ' ', text)
    # Convert to lowercase
    text = text.lower()
    # Remove punctuation and special characters
    text = re.sub(r'[^a-zA-Z0-9\s]', '', text)
    # Remove stop words
    stop_words = set(stopwords.words('english'))
    words = text.split()
    text = ' '.join([word for word in words if word not in stop_words])
    return text

df['preprocessed_review'] = df['review'].apply(preprocess_text)
display(df[['review', 'preprocessed_review']].head())

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


review  \
0  One of the other reviewers has mentioned that ...   
1  A wonderful little production. <br /><br />The...   
2  I thought this was a wonderful way to spend ti...   
3  Basically there's a family where a little boy ...   
4  Petter Mattei's "Love in the Time of Money" is...   

                                 preprocessed_review  
0  one reviewers mentioned watching 1 oz episode ...  
1  wonderful little production filming technique ...  
2  thought wonderful way spend time hot summer we...  
3  basically theres family little boy jake thinks...  
4  petter matteis love time money visually stunni...

## Word count and vocabulary (python)

### Subtask:
Calculate the total number of words and unique words in the corpus using pure Python.


**Reasoning**:
Calculate the total number of words and unique words in the corpus using pure Python by iterating through the preprocessed reviews.



In [4]:
all_words = set()
unique_words = set()

for review in df['preprocessed_review']:
    words = review.split()
    all_words.update(words)
    unique_words.update(words)

total_words_count = len(all_words)
unique_words_count = len(unique_words)

print(f"Total number of words in the corpus: {total_words_count}")
print(f"Number of unique words in the corpus: {unique_words_count}")

Total number of words in the corpus: 154351
Number of unique words in the corpus: 154351


## One-hot encoding

### Subtask:
Apply One-Hot Encoding to a sample of the preprocessed text data.


**Reasoning**:
Import OneHotEncoder, create a sample of the preprocessed data, reshape it, instantiate the encoder, fit and transform the data, and display the shape of the result.



In [5]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np

sample_size = 1000
sampled_data = df['preprocessed_review'].sample(sample_size, random_state=42).values.reshape(-1, 1)

encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
one_hot_encoded_data = encoder.fit_transform(sampled_data)

print(f"Shape of the one-hot encoded data: {one_hot_encoded_data.shape}")

Shape of the one-hot encoded data: (1000, 999)


## Bag of words

### Subtask:
Apply Bag of Words to the preprocessed text data and find the vocabulary and word counts.


**Reasoning**:
Apply Bag of Words to the preprocessed text data and find the vocabulary and word counts.



In [6]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
bag_of_words_matrix = vectorizer.fit_transform(df['preprocessed_review'])

vocabulary = vectorizer.get_feature_names_out()
word_counts = bag_of_words_matrix

print(f"Shape of the Bag of Words matrix: {bag_of_words_matrix.shape}")

Shape of the Bag of Words matrix: (43525, 154323)


## Bag of bigrams and trigrams

### Subtask:
Apply Bag of Bigrams and Trigrams to the preprocessed text data and observe the dimensionality of the vocabulary.


**Reasoning**:
Apply Bag of Bigrams and Trigrams using CountVectorizer with ngram_range=(2, 3) and print the shape of the resulting matrix.



In [7]:
from sklearn.feature_extraction.text import CountVectorizer

bigram_trigram_vectorizer = CountVectorizer(ngram_range=(2, 3))
bigram_trigram_matrix = bigram_trigram_vectorizer.fit_transform(df['preprocessed_review'])

print(f"Shape of the Bag of Bigrams and Trigrams matrix: {bigram_trigram_matrix.shape}")

Shape of the Bag of Bigrams and Trigrams matrix: (43525, 7688033)


## Tf-idf

### Subtask:
Apply TF-IDF to the preprocessed text data, calculate IDF scores, and find the vocabulary.


**Reasoning**:
Apply TF-IDF to the preprocessed text data, calculate IDF scores, and find the vocabulary.



In [8]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(df['preprocessed_review'])

vocabulary_tfidf = tfidf_vectorizer.get_feature_names_out()
idf_scores = tfidf_vectorizer.idf_

print(f"Shape of the TF-IDF matrix: {tfidf_matrix.shape}")

Shape of the TF-IDF matrix: (43525, 154323)


## Summary:

### Data Analysis Key Findings

*   The dataset contains 43,525 reviews and their corresponding sentiments.
*   Preprocessing steps, including removing HTML tags, converting to lowercase, removing punctuation and special characters, and removing stop words, were successfully applied to the review text.
*   Using a pure Python approach, the corpus was found to have a total of 154,351 words and 154,351 unique words in the preprocessed text.
*   Applying One-Hot Encoding to a sample of 1000 preprocessed reviews resulted in a matrix shape of (1000, 999), indicating 999 unique words in that sample.
*   The Bag of Words model applied to the entire preprocessed corpus resulted in a document-term matrix with a shape of (43525, 154323), indicating a vocabulary size of 154,323 unique words.
*   Applying the Bag of Bigrams and Trigrams model resulted in a matrix with a shape of (43525, 7688033), highlighting a significantly larger vocabulary when considering sequences of 2 or 3 words.
*   TF-IDF applied to the preprocessed corpus also resulted in a matrix shape of (43525, 154323), with a vocabulary size of 154,323 unique words. IDF scores were also calculated.

### Insights or Next Steps

*   The significant difference in vocabulary size between Bag of Words (unigrams) and Bag of Bigrams/Trigrams suggests that considering word sequences captures much richer linguistic information, albeit at the cost of much higher dimensionality.
*   The next steps could involve using these different vectorization techniques (Bag of Words, Bag of Bigrams/Trigrams, or TF-IDF) as features for training a sentiment classification model and comparing their performance.
